cargo-guppy/src/lib.rs

```diff
@@ -18,18 +25,136 @@ pub fn cmd_diff(json: bool, old: &str, new: &str) -> Result<(), Error> {
 
     Ok(())
 }
 
-pub fn cmd_dups() -> Result<(), Error> {
+pub fn cmd_dups() -> Result<(), anyhow::Error> {
     let lockfile = Lockfile::from_file("Cargo.lock")?;
 
     lockfile.duplicate_packages();
 
     Ok(())
 }
+
+arg_enum! {
+    #[derive(Debug)]
+    pub enum Kind {
+        All,
+        Workspace,
+        ThirdParty,
+    }
+}
+
+struct NameVisitor;
+
+impl PackageDotVisitor for NameVisitor {
+    fn visit_package(&self, package: &PackageMetadata, mut f: DotWrite<'_, '_>) -> fmt::Result {
+        write!(f, "{}", package.name())
+    }
+
+    fn visit_link(&self, _link: DependencyLink<'_>, mut f: DotWrite<'_, '_>) -> fmt::Result {
+        write!(f, "")
+    }
+}
+
+pub fn cmd_select(
+    kind: Kind,
+    target: Option<String>,
+    output: Option<String>,
+    include_dev: bool,
+    include_build: bool,
+    omit_edges_into: &[String],
+    roots: &[String],
+) -> Result<(), anyhow::Error> {
+    let mut pkg_graph = PackageGraph::default();
+    let mut package_ids = HashSet::new();
+    let mut omitted_package_ids = HashSet::new();
+
+    let root_set: HashSet<String> = if roots.len() > 0 {
+        roots.iter().cloned().collect()
+    } else {
+        pkg_graph
+            .packages()
+            .filter_map(|meta| {
+                if meta.in_workspace() {
+                    Some(meta.name().to_string())
+                } else {
+                    None
+                }
+            })
+            .collect()
+    };
+
+    let omitted_set: HashSet<String> = omit_edges_into.iter().cloned().collect();
+
+    for metadata in pkg_graph.packages() {
+        if root_set.contains(metadata.name()) {
+            package_ids.insert(metadata.id().clone());
+        }
+        if omitted_set.contains(metadata.name()) {
+            omitted_package_ids.insert(metadata.id().clone());
+        }
+    }
+
+    let transitive_deps: Vec<_> = pkg_graph
+        .select_forward(&package_ids)?
+        .into_iter_ids(None)
+        .chain(&package_ids)
+        .cloned()
+        .collect();
+
+    pkg_graph.retain_edges(|_, DependencyLink { from, to, edge }| {
+        let include_kind = match kind {
+            Kind::All | Kind::ThirdParty => true,
+            Kind::Workspace => from.in_workspace() && to.in_workspace(),
+        };
+
+        let include_target = if let Some(ref target) = target {
+            edge.normal()
+                .and_then(|meta| meta.target())
+                .and_then(|edge_target| {
+                    let res = target_spec::eval(edge_target, target).unwrap_or(true);
+                    Some(res)
+                })
+                .unwrap_or(true)
+        } else {
+            true
+        };
+
+        let include_type = edge.normal().is_some()
+            || include_dev && edge.dev().is_some()
+            || include_build && edge.build().is_some();
+
+        let include_dep = transitive_deps.contains(from.id()) && transitive_deps.contains(to.id());
```

Comment:
So I've been staring at this for a bit -- I'm trying to follow some of the logic here and I'm having trouble following this. Could you explain what the `include_dep` filter is trying to do?

Reply:
I'll add comments to all the filtering lines so they are more clear. This one filters out edge to/from packages that are in the metadata (ie, all packages from Cargo.lock) but not in the transitive closure from the given roots.

Reply:
Thanks for the comments! They're super helpful.

Ahh interesting, do you actually need that filtering? Because I think the line at:

```
for package in pkg_graph.select_forward(&package_ids)?.into_iter_ids(None) { 
```

will end up achieving the same result anyway. I might totally be missing something here though!

Reply:
You're right. Now that everything after retain_edges uses select_forward, the filtering happens automatically.